### DISCUSSION TITLE
Support for EXE/DLL embedded resources

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#238] [General]
As of v0.10.2773, twinBASIC now supports embedding of resource data and files into the generated EXEs and DLLs.   The twinBASIC virtual file system is used to categorize the resources, and so we won't be needing a separate 'resource editor' like we had in VB6.

We have full support for BITMAP, ICON and CURSOR formats, where the linker has to massage them into their resource equivalent formats at compile time.   To use these, create a new folder called either BITMAP / ICON / CURSOR in your projects Resources folder, and you can drag and drop files from Windows/Explorer into the new folder.

<img width="225" alt="tb_resources1" src="https://user-images.githubusercontent.com/25738109/123414674-19ddf580-d5ac-11eb-802f-0366775f735a.png">

For those that are familiar with EXE resources, you can identify resources either by a text name, or by an ordinal.  To create an ordinal based folder or resource, rename it starting with #, e.g. `#101` will create resource with the 101 identifier.

Of course this update would not be complete without the corresponding library support.   So we've also got support for the VB6 functions LoadResData and LoadResPicture, which are currently implemented in the _HiddenModule module, but will later be exposed via the VB Global class once we've added that to match VB6.  LoadResPicture returns an IPictureDisp object, which is handy for putting icons in menus and ribbons.  LoadResData returns a byte-array containing the raw resource contents.

The samples have been updated to include the tB logo in the resources ICON folder, so that Windows/Explorer now shows the tB icon...
<img width="242" alt="tb_Sample1Update" src="https://user-images.githubusercontent.com/25738109/123414315-a4722500-d5ab-11eb-93e9-004967914e94.png">

The Sample 5 project has been updated to show how to create a ribbon with an embedded image for the button.
<img width="533" alt="tb_Sample5Update" src="https://user-images.githubusercontent.com/25738109/123413554-d3d46200-d5aa-11eb-93bc-8e7912b21c93.png">
<img width="433" alt="tb_Sample5Update2" src="https://user-images.githubusercontent.com/25738109/123413787-10a05900-d5ab-11eb-8c1a-0ed6a87086cf.png">

_In a future revision, we will be able to add VB6 .res files into the project, and the contents will be merged into the generated EXEs/DLLs._

**_Support for manifest resources and version information resources will be coming in the next few days_**

**Comment by WaynePhillipsEA** [comment]
Resources are particularly helpful when it comes to needing language/locale specific resources.  These are also supported in twinBASIC (though string tables are not... yet).  

To create a language/locale specific resource, create a subfolder under the main folder type named 'LCID_xxxx' where xxxx represents the 4-digit locale ID in hex format.  For a list of LCIDs, see https://www.voidtools.com/support/everything/language_ids/    

For example, resource files in the folder `Resources\BITMAP\LCID_0809\` would override resources in the `Resources\BITMAP\` folder, on a file-by-file basis, if your system is set to use the 0x0809 LCID (United Kingdom).  This feature is currently not working when running inside the IDE, where the LCID specific resources are being ignored.

**Comment by Kr00l** [comment]
I guess strings (LoadResString) will be supported soon as well ?

**Reply by WaynePhillipsEA** [reply]
@Kr00l Yes, I need to create a little editor to allow creating them, and then I'll implement LoadResString.  At some point next week, they should be supported.

**Comment by Kr00l** [comment]
Small **cosmetic** thing with the great manifest creatior.
The assemblyIdentify is done differently, see screenshot.
![image](https://user-images.githubusercontent.com/33756666/123595743-8dbd0f80-d7f1-11eb-8fe5-1f7e70528305.png)

Also it is legal to have` processorArchitecture="*"` instead of `processorArchitecture="x86"`
I think this makes it easier when converting a project to x64. (?)

**Reply by WaynePhillipsEA** [reply]
Thanks!   Actually the processorArchitecture field looks to be optional, so it might be best to just remove that field.

**Reply by WaynePhillipsEA** [reply]
Hmmm... according to https://docs.microsoft.com/en-us/windows/win32/sbscs/application-manifests, the processorArchitecture field is optional, but an EXE built with a manifest that doesn't specify the processorArchitecture doesn't work.   So you're quite right... "*" is the way to go.

**Reply by WaynePhillipsEA** [reply]
Sorted in v0.10.2822

**Reply by Kr00l** [reply]
> Hmmm... according to https://docs.microsoft.com/en-us/windows/win32/sbscs/application-manifests, the processorArchitecture field is optional, but an EXE built with a manifest that doesn't specify the processorArchitecture doesn't work. So you're quite right... "*" is the way to go.

Typical confusion by MS. On the page [Enabling Visual Styles](https://docs.microsoft.com/en-us/windows/win32/controls/cookbook-overview) they state that "x86" or "amd64" should be used. But they recommend on that page to use "*".

Anyhow it's sorted out. Thanks for the update.

**Comment by Kr00l** [comment]
In v0.10.2822 all sample projects were updated with the default manifest, which is great.

**But** there are two types of manifest id's.
1 = Process manifests with ID CREATEPROCESS_MANIFEST_RESOURCE_ID (1), used during process creation
2 = Isolation-Aware manifests, with ID ISOLATIONAWARE_MANIFEST_RESOURCE_ID (2), used during DLL loading

So, for the ActiveX DLL sample project there are two possibilities.
Either remove the default manifest. So the executable manifest is used for the visual styles.
Or change the ID from 1 to 2 for the ActiveX DLL sample project.

**Reply by WaynePhillipsEA** [reply]
Oh you're quite right, I don't know what I was thinking!   I'll get that fixed.   We can't use the isolation aware version because that requires some work at all entry points (and needs a flag to be set in the PE headers).  So I'll remove the manifest from the DLL samples for now, as that is definitely wrong.

**Reply by WaynePhillipsEA** [reply]
Removed in v0.10.2914, thanks

**Comment by WaynePhillipsEA** [comment]
In v0.10.2914, we now support string tables for localization.  Currently, the support is via raw JSON files, but there will be an editor coming later.  Right click on Resources folder...

<img width="355" alt="tb_createStringResource" src="https://user-images.githubusercontent.com/25738109/123764903-24a2ce00-d8bd-11eb-8547-1ef7781b01c1.png">

<img width="463" alt="tb_createStringResource2" src="https://user-images.githubusercontent.com/25738109/123764914-28365500-d8bd-11eb-934a-62ba8ae5f799.png">

It is a good idea to include a [name] field, so that when #242 is implemented, the names can be used to access the strings via the object model (e.g. Resources.Strings.MyLocalizedString1).  At the moment, you'll need to use the legacy VB6 compatible function LoadResString(id)

**Reply by WaynePhillipsEA** [reply]
You can create multiple JSON files in the STRING resources folder, and the linker will merge them, allowing you to separate and categorize your strings into groups.

**Comment by Kr00l** [comment]
The ICON is not displayed anymore like in a previous version of tB.

![image](https://user-images.githubusercontent.com/33756666/123765765-54a29f00-d8c6-11eb-8cfa-d2789ac249cf.png)

**Reply by WaynePhillipsEA** [reply]
@Kr00l you can rename the file to have an extension of .ico, e.g. `#101.ico`, and then VS code should show the icon.  The linker/resource-builder will ignore the extension.

**Reply by Kr00l** [reply]
Worked. Thx

**Comment by Kr00l** [comment]
Only version information block still missing?
